odoo/addons/test_read_group/tests/test_m2m_grouping.py

```diff
@@ -216,6 +216,26 @@ def test_complex_case(self):
             },
         ])
 
+    def test_order_by_many2one_id(self):
+        # ordering by a many2one ordered itself by id does not use useless join
+        expected_query = '''
+            SELECT
+              min("test_read_group_order_line".id) AS id,
+              count("test_read_group_order_line".id) AS "order_id_count",
+              "test_read_group_order_line"."order_id" as "order_id"
+            FROM "test_read_group_order_line"
+            GROUP BY "test_read_group_order_line"."order_id", "test_read_group_order_line"."order_id"
```

Comment:
It is weird to repeat the groupby clause (`"test_read_group_order_line"."order_id"`). I think you should move your if condition in `_read_group_prepare` just after `if self._fields[order_field.split(':')[0]].type == 'many2one':`.

(sry I didn't see it before)

Reply:
without the fix, there was already an unnecessary groupby, for example:

```sql
  GROUP BY "test_read_group_order_line"."order_id", "test_read_group_order_line__order_id"."id"
  ORDER BY "test_read_group_order_line__order_id"."id"
```

that useless groupby always happen even with an many2one model not ordered by ID, maybe we could remove it in all case:

```diff
                     if order_clause:
                         orderby_terms.append(order_clause)
+                        groupby_terms.remove(next(
+                            gb['qualified_field'] for gb in annotated_groupbys
+                            if gb['field'] == order_field
+                        ))
                         groupby_terms += [order_term.split()[0] for order_term in order_clause.split(',')]
```
or just for the use case where it's a duplicate (but that would only be my case of "group by m2o and order by the same m2o that is ordered by ID"):

```diff
                     if order_clause:
                         orderby_terms.append(order_clause)
-                        groupby_terms += [order_term.split()[0] for order_term in order_clause.split(',')]
+                        add_groupby_terms = [order_term.split()[0] for order_term in order_clause.split(',')]
+                        groupby_terms = [term for term in groupby_terms if term not in add_groupby_terms] + add_groupby_terms
```
for doing the fix in this part, it was rather ugly (and I had not already hadded the DESC/ASC/...) and this would leave the `_generate_order_by` inside `_search` so it doesn't seem too nice to me but I can check it further.

Reply:
I was thinking about modify `_read_group_prepare` like this:
```python
                if self._fields[order_field.split(':')[0]].type == 'many2one':
```
into => 
```python
                field = self._fields[order_field.split(':')[0]]
                if field.type == 'many2one' and self.env[field.comodel_name]._order != 'id':
```
And keep the change in `_generate_m2o_order_by` too for the search method (and add a test for it).
It avoids repeating the groupby and cover the case of the search too. It is uglier, but I don't see another solution to have queries that makes sense.

Also, careful about the FW branches, I refactored '_read_group' in 16.3/16.4. At this point, can you move your test in the file called 'private_read_group_test.py'.

Reply:
I applied it, this solve the case where the groupby is duplicate exactly (and this changes the "order by" to a non-qualified field, but it was already the case for non-m2o fiels)